### PR TITLE
Labelers update

### PR DESCRIPTION
**Pull request by guolin1**
Labelers update:

`InpatientReadmissionLabeler` : exclude same-day readmissions since prediction time is by-default shifted to the end of the discharge day - adopted from [here](https://github.com/som-shahlab/femr/blob/6559ec50c275008cac60a471c07d73f848a944b6/src/piton/labelers/omop_inpatient_admissions.py#L113). 
`InpatientLongAdmissionLabeler`: exclude if discharge (for the same admission) or death occurred before prediction time
`InpatientMortalityLabeler`: exclude if discharge (for the same admission) or death occurred before prediction time
`ThrombocytopeniaLabValueLabeler`: add LOINC/777-3 to concept codes
`HyperkalemiaLabValueLabeler` add LOINC/6298-4 and LOINC/2823-3 to concept codes
`HypoglycemiaLabValueLabeler`: add LOINC/14749-6 and LOINC/15074-8 to concept codes
`HyponatremiaLabValueLabeler`: add LOINC/2951-2, and LOINC/2947-0 to concept codes
`AnemiaLabValueLabeler`:  add unit conversion for "g/L" 

Others:
`map_omop_concept_codes_to_femr_codes` issues warning instead of raising an exception when concept code is not referenced in the patient database. 
`compute_random_num`: allows 64-bit int

### CODE REVIEW

File: src/femr/labelers/omop_inpatient_admissions.py [hunk 1]
```diff
@@ -143,11 +148,28 @@ def __init__(
...
             prediction_time: datetime.datetime = self.prediction_time_adjustment_func(admission_time)
+
+            # exclude if discharge or death occurred before prediction time
+            death_concepts: Set[str] = map_omop_concept_codes_to_femr_codes(self.ontology, get_death_concepts())
+            death_times: List[datetime.datetime] = []
+            for e in patient.events:
+                if e.code in death_concepts:
+                    death_times.append(e.start)
+
+            if prediction_time > discharge_time:
```
Comment (Lalaland) [inline src/femr/labelers/omop_inpatient_admissions.py:167]: Isn't this duplicative with the modifications to get prediction times?

File: src/femr/labelers/omop_inpatient_admissions.py [hunk 2]
```diff
@@ -143,11 +148,28 @@ def __init__(
         self.prediction_time_adjustment_func = prediction_time_adjustment_func
 
     def label(self, patient: Patient) -> List[Label]:
-        """Label all admissions with admission length > `self.long_time`"""
+        """
+        Label all admissions with admission length > `self.long_time`
+        Exclude admission if time of discharge or death < prediction time
```
Comment (Lalaland) [inline src/femr/labelers/omop_inpatient_admissions.py:153]: This is a neat change. I think we might actually want to make this universal (make it so that all predictions after death are discarded). Would be another PR, but what would your thoughts be about that?

**Review by Lalaland — CHANGES_REQUESTED** [review]
Looks mostly good with two minor comments! Thanks for the fixes.

File: src/femr/labelers/omop_inpatient_admissions.py (hunk 2 above)
Comment (guolin1) [reply src/femr/labelers/omop_inpatient_admissions.py:153]: That is a good idea. I can't think of a scenario for which we wouldn't apply this rule.

File: src/femr/labelers/core.py [hunk 3]
```diff
@@ -556,7 +556,7 @@ def get_labeler_type(self) -> LabelType:
 
 
 def compute_random_num(seed: int, num_1: int, num_2: int):
```
[the remark this answers is not in the document]
Comment (guolin1) [reply src/femr/labelers/core.py:558]: Will do!

File: src/femr/labelers/omop_inpatient_admissions.py [hunk 4]
```diff
@@ -109,9 +109,14 @@ def get_outcome_times(self, patient: Patient) -> List[datetime.datetime]:
     def get_prediction_times(self, patient: Patient) -> List[datetime.datetime]:
         """Return end of admission as prediction timm."""
         times: List[datetime.datetime] = []
-        for __, discharge_time in get_inpatient_admission_discharge_times(patient, self.ontology):
+        admission_times = set()
+        for admission_time, discharge_time in get_inpatient_admission_discharge_times(patient, self.ontology):
             prediction_time: datetime.datetime = self.prediction_time_adjustment_func(discharge_time)
+            # Ignore patients who are readmitted the same day they were discharged b/c of data leakage
+            if prediction_time.replace(hour=0, minute=0, second=0, microsecond=0) in admission_times:
```
[the remark this answers is not in the document]
Comment (guolin1) [reply src/femr/labelers/omop_inpatient_admissions.py]: I think the previous logic might be wrong. Just updated with a simple way of checking whether current admission date <= previous discharge date (including "<" just in case there are overlapping visit events).

File: src/femr/labelers/omop_inpatient_admissions.py (hunk 1 above)
Comment (guolin1) [reply src/femr/labelers/omop_inpatient_admissions.py:170]: I don't think so. The prediction time is the adjusted admission time, and this checks whether the adjustment has moved the prediction time to > discharge time, and if so, exclude.

Comment (guolin1) [reply src/femr/labelers/omop_inpatient_admissions.py:170]: Actually, I take that back... what I just changed is incorrect. Will think it through and consider your suggestion.

File: src/femr/labelers/core.py [hunk 5]
```diff
@@ -556,9 +556,9 @@ def get_labeler_type(self) -> LabelType:
...
-    network_num_1 = struct.pack("!I", num_1)
-    network_num_2 = struct.pack("!I", num_2)
-    network_seed = struct.pack("!I", seed)
+    network_num_1 = struct.pack("!Q", num_1)
```
Comment (Lalaland) [inline src/femr/labelers/core.py:559]: This one should be lowercase q as well.

File: src/femr/labelers/omop_inpatient_admissions.py [hunk 6]
```diff
@@ -109,9 +109,17 @@ def get_outcome_times(self, patient: Patient) -> List[datetime.datetime]:
...
         """Return end of admission as prediction timm."""
         times: List[datetime.datetime] = []
-        for __, discharge_time in get_inpatient_admission_discharge_times(patient, self.ontology):
+        prev_discharge_date: datetime.date = datetime.date(1900, 1, 1)
+
+        for admission_time, discharge_time in get_inpatient_admission_discharge_times(patient, self.ontology):
             prediction_time: datetime.datetime = self.prediction_time_adjustment_func(discharge_time)
+
+            # Ignore patients who are readmitted the same day they were discharged b/c of data leakage
+            if admission_time.date() <= prev_discharge_date:
```
Comment (Lalaland) [inline src/femr/labelers/omop_inpatient_admissions.py:118]: This might seem stupid, but which isn't this just 

```
if prediction_time >= discharge_time:
    continue
```

**Review by Lalaland — CHANGES_REQUESTED** [review]
Might be good to chat about the logic here after our Wednesday meeting? I'm still confused why you can't simply compare prediction time to discharge time instead of tracking the prior admission.

File: src/femr/labelers/omop_inpatient_admissions.py (hunk 6 above)
Comment (guolin1) [reply src/femr/labelers/omop_inpatient_admissions.py:118]: For readmission predictions, the prediction time is by default 11:59 PM on discharge date, and so it should always be >= discharge_time. Instead, we want to exclude any re-admissions that occurred on the same date as discharge, since these re-admissions would've occurred before or on prediction time.

Comment (Lalaland) [reply src/femr/labelers/omop_inpatient_admissions.py:118]: Oh, I see. Good catch and thanks for finding it. This should also probably be made general across all labelers, just not in this PR.

**Comment by Lalaland** [comment]
I don't think we need to chat then, as that makes things much more clear. We do probably want to fix these unit tests before merging. Let me know if you want me to do that, they are done in a confusing manner to a big extent.

**Comment by guolin1** [comment]
Thanks @Lalaland! Yeah would be great if you could help with fixing the unit tests.

File: src/femr/labelers/omop_inpatient_admissions.py [hunk 7]
```diff
@@ -185,3 +210,46 @@ def get_outcome_times(self, patient: Patient) -> List[datetime.datetime]:
...
+        ]
+        prediction_end_times: List[datetime.datetime] = [self.visit_end_adjust_func(visit.end) for visit in visits]
+        outcome_times: List[datetime.datetime] = self.get_outcome_times(patient)
+        death_time = min(outcome_times) if outcome_times else None
+        results: List[Label] = []
+
+        # For each visit, check if there is an outcome which occurs within the (start, end) of the visit
+        for prediction_start, prediction_end in zip(prediction_start_times, prediction_end_times):
+            # exclude if prediction time > discharge time
+            if prediction_start >= prediction_end:
```
Comment (Lalaland) [inline src/femr/labelers/omop_inpatient_admissions.py:231]: @guolin1 I made a swap here from ">" to ">=" because I think that's the more correct logic here. Can you double check I did this right?

The idea is that if prediction_start == prediction_end, then it's too easy / trivial of a prediction.

**Comment by Lalaland** [comment]
Ok, fixed the tests. @guolin1 once you check the one other minor change that I made, I think this is ready to merge.

File: src/femr/labelers/omop_inpatient_admissions.py (hunk 7 above)
Comment (guolin1) [reply src/femr/labelers/omop_inpatient_admissions.py:231]: That makes sense. Looks good to me! Thanks!

**Review by Lalaland — APPROVED** [review]

**Comment by Lalaland** [comment]
@guolin1 Feel free to merge this in whenever you want!
